### DISCUSSION TITLE
Reimplement the ability to set a base slider velocity multiplier in setup

### DISCUSSION THREAD

**Discussion by furiner** [ppy/osu#23391] [Ideas] [answered]
Firstly, I'd like to mention that I acknowledge that this has been discussed previously in #11003 as not planned; but I'd like to offer a reasoning as to why I believe it would be a good idea to reimplement.

Right now, for legacy purposes; the `BeatmapDifficulty` realm object already still retains the `SliderMultiplier` parameter that can be found in a .osu file, and it is profoundly simple to reimplement it into the setup editor screen, [as I have conceptually done here].(https://cdn.discordapp.com/attachments/234057206784458754/1103299912982077541/osu_xRpfcjQJgh.png)

As it stands, I believe that re-adding the ability to change the base slider velocity in the beatmap editor would be valuable for ease of use. The current system is quite different from the one used in osu! stable, which has been used by many mappers for a long time. Mappers are used to using relative values for slider velocity in proportion to the base velocity assigned to a map. Without the ability to customize the base velocity, it becomes more difficult to modify values used in the past. This would require mappers to readjust their values and spend much more time experimenting to achieve the desired velocity, which could be simplified with the existence of a base velocity.

In addition to ease of use, having a base slider velocity also makes it easier to give and receive feedback on beatmaps. Since most mappers are used to working with relative values in proportion to the base velocity, it's easier to understand how the changes to slider velocity affect the beatmap, as they're more foreseeably drastic in nature. This is (at least in my eyes) important for receiving feedback on a beatmap's pacing and difficulty, and easily ensuring that it meets the standards set forth in the osu! ranking criteria.

Furthermore; I'd like to pitch in with the fact that I tend to find myself going back and forth from osu! lazer and stable solely because of this option, because I and others highly prefer the ease of such shorter values. I believe this would be a great move for improving the ability for mappers to move from stable to lazer, because the removal of such a feature comes as a relatively gigantic change to how they view slider velocity, even if it may not look like it.

**Comment by peppy** [comment]
I understand what you're saying, but what stops a mapper from setting a base velocity of 1x and then using 4x everywhere in the beatmap?

I think this would need to be addressed from two angles:

- There should be a way to easily see how much of each SV is used in a beatmap (maybe at the setup screen there would be summary details showing the mode SV along with most commonly used in order of usage?)
- The last used velocity should be saved somewhere so when returning to the editor you have that selected.. maybe? We already use the SV of the closest hit object which does work quite well so I'm a bit iffy on this.

You've included a lot of text but I still fail to understand exactly what you are missing here. You are implying there's a "base" SV, but does that mean you want all the per-object SVs to be relative to that base value? ie. set the beatmap to 4x, which means on the timeline a 1x velocity is actually 4x? (seems confusing to me)

Maybe you can make a video showing the troubles you are encountering?

**Reply by furiner** [reply]
`I understand what you're saying, but what stops a mapper from setting a base velocity of 1x and then using 4x everywhere in the beatmap?`

This is true, but typically (at least in regards to stable) most people won't generally do this as they are typically already used to the fact the slider multiplier value can be set as per map. When I say "base SV", I am referring to the option you are given in the timing menu on stable.

![osu!_JrYVcc4O1d](https://user-images.githubusercontent.com/30084105/236385911-88a775e0-d2ab-4c5c-be55-f480f4e2c3c4.png)

I am arguing that this could be easily reimplemented as a configurable option for the ease of of the points I have listed above, and ultimately nothing else would need to change as Lazer already properly handles it due to its existence in the .osu file in coordination with stable. If people wish to still use 1.0x velocities, then they would be fine to do so; but my primary point at hand is for those who greatly would prefer the existence of this due to the ease of remote values that people typically set via. uninherited timing points set that are handled relatively by osu! in proportion to the base velocity assigned to that map they are working on in particular.

At least in my eyes, I and others find it generally easier to pinpoint where I want the velocity of all the sliders in a map in proportion to the intensity of a map to place; typically setting the base value within a range of 1.8x to 2.4x in proportion to a song's BPM or overall intensity.

**Reply by peppy** [reply]
So, lazer already chooses a SV based on where you are in the beatmap, based on nearby hitobjects, right? If this was unchanged, the only time the "Beatmap Setup" SV would be used is on placing the first hitobject ever. Sounds weird.

It sounds like you are most interested in seeing the common velocity, so what would you say if as I proposed above, the mode SV was displayed alongside the SV controls:

![osu! 2023-05-05 at 06 17 07](https://user-images.githubusercontent.com/191335/236388772-d5676f93-98bf-4412-9406-a80ab6f3049a.png)

**Reply by furiner** [reply]
While that does sound like an intriguing feature to see, especially for consistency and linearity; I feel the only approach that would solve is to maintain a consistent velocity across the map; which I do see where my points may have been interpreted in that regard.

Here, let me name a few circumstances as to what I am trying to get across.

In this map, I set a base velocity of 2.2x on a 151bpm song because I want it's intensity to properly represent that of a map that requires you to alternate keys (alt mapping) and a song that has a lot of use-case for such a high base velocity. I felt that 2.2x would get that across easily, whilst also not being too fast for such a slow song.
![osu!_6jArimikJX](https://user-images.githubusercontent.com/30084105/236389337-8c864197-4816-43f6-b535-d7c40fa6cc64.png)

And so, the sliders I use will be in proportion to that velocity; where I can use simpler values to figure out where I want a sliders speed; rather than directly toying around with the velocity unless I feel the velocity I am using for a slider is too fast. This is for me a part of the intuition that comes as a mapper, and I find it very difficulty to stray away from it.
![osu!_fDMt03va8O](https://user-images.githubusercontent.com/30084105/236389568-9295bd38-2f52-4e01-a86d-3afc7fee4326.png)

Here, I use 1.9x on a 210 bpm song that's more full of streams, so I don't need a very high base velocity to represent a song that isn't very intense.
![osu!_5QBB5gdIyX](https://user-images.githubusercontent.com/30084105/236389841-0bfb4626-6094-4da8-9521-fd06759b7272.png)

Typically, under these circumstances; the sliders I use will be much closer to 1.0x rather than anything between 1.4 or 2.0x as shown above in the map I wanted to be more representative of the intensity it has as a song. However, most of this short linearity is left to streams so as I won't be using values that would make it difficulty for a player to aim the kick sliders properly. The parts that do not have streams have velocities faster in nature in proportion to the intensity of that part of the song.
![osu!_jPr9l6bUVb](https://user-images.githubusercontent.com/30084105/236390072-0bab684d-fa10-46ac-be85-277154aaa8d7.png)

What I am trying to get across is that I use base velocity to set the baseline for where I want the general overall intensity of a map to be, and use the relative velocity per object to represent the intensity of that section.

I will clarify that the option here is one that I coded in myself, as it was very easy to add; so it is my example.

**Reply by peppy** [reply]
So it's just a slider to convey mapper notes? I don't think that's going to happen, sorry.

**Reply by furiner** [reply]
Is that not a valid form of mapper expression? I fully believe that the notes and inner details as to what a map is trying to convey is just as valid as anything else. I believe that the editor should allow the mapper to convey what they wish to output through their maps, and this is one of the means people have at doing so.

It's also relatively extremely simple to add, and I don't see any reason it should have ever been removed to begin with.

**Reply by peppy** [reply]
Will wait for more feedback on this, but I'm generally against having settings which have zero effect. It'll only add to confusion.

**Reply by maotovisk** [reply]
I'd say they have a valid point, there are some really useful use-cases for a non-1x-base-sv:

1. Determining the general pacing of the map, specially for lower diffs;
2. More ease-of-map for grid-based mapping (with 1,92x being the most common value for 1/1 and 1/2 sliders to snap to the biggest grid size);
3. It's something that's already into the mapping culture, as a general setting that mappers often use even to take as inspiration from other mappers.

**Reply by peppy** [reply]
@maotovisk have you used the lazer editor? do you first understand that there is no global SV setting anymore?

have you checked the linked PR i made where it will show map-wide stats on SV, which basically replicates the "map SV"?

do you understand even if it was added back, it would *not actually be used anywhere else other than the "setup" tab*?

**Reply by maotovisk** [reply]
Yes I have, but if I did not read it wrongly, the entire OP point is to bring back the ability of setting the base value, since it's already stored and supported. 

In relation to your first answer:
> which means on the timeline a 1x velocity is actually 4x? (seems confusing to me)

Yes. That's exactly the point.

> do you understand even if it was added back, it would not actually be used anywhere else other than the "setup" tab?

I think it is fine, if it worked the like mentioned above, as a global setting per difficulty.

**Reply by peppy** [reply]
Okay, turns out I was wrong and there *is* a `SliderMultiplier` which behaves as I mentioned above (premultiplied, becomes the "1.0x" for the beatmap).

So we'll may want to add back this basic behaviour, but also of note, it's not currently implemented in a way it will update in the editor when modified. And adding it back is going to make displaying things per hitobject confusing in one way or another (either we always show the post-multiplied as the adjustable value, or have it shown below. and i don't even know what beatmap wide numbers should display anymore).

**Reply by furiner** [reply]
At least from adding it back manually on my own it did update the same way other options would have, and it made it function the exact same way it did in stable. However; I will admit that showing the post-multiplied value in the side inspector would be an outstanding compromise for readding the feature; because it resolves the issues that I mentioned in the past while retaining the feature I've felt was long missing.

**Reply by furiner** [reply]
I'll probably attempt to make a PR directly for easy testing and tracking for the feature.

**Reply by peppy** [reply]
I already have a branch making the change, will PR soon.

**Reply by peppy** [reply, marked as answer]
#23431

**Comment by Walavouchey** [comment]
what base slider velocity allows is setting a default value for every slider, such that you don't have to always change the sv of every placed slider to have it in the ballpark of what's expected for some particular difficulty

but given that lazer "already uses the SV of the closest hit object", there shouldn't be any other problem than the numerical sv values being different to what mappers are used to, which would be a weird thing to cater to anyway

**Reply by furiner** [reply]
Wouldn't it be easier to at least retain the setting for the vast portion of users who *do* use that setting within good purpose rather than having to press them to inevitably accommodate in a way that can come off unintuitively?
